src/Runner.Worker/JobExtension.cs

```diff
@@ -417,6 +417,43 @@ public void FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRequestM
                         }
                     }
 
+                    // Evaluate environment data
+                    if (jobContext.ActionsEnvironment?.Url != null)
+                    {
+                        try
+                        {
+                            context.Output($"Evaluate and set environment url");
+
+                            // Populate env context for each step
+                            Trace.Info("Initialize Env context for evaluating environment url");
+#if OS_WINDOWS
+                            var envContext = new DictionaryContextData();
+#else
+                            var envContext = new CaseSensitiveDictionaryContextData();
+#endif
+                            context.ExpressionValues["env"] = envContext;
+                            foreach (var pair in context.Global.EnvironmentVariables)
+                            {
+                                envContext[pair.Key] = new StringContextData(pair.Value ?? string.Empty);
+                            }
+
+                            Trace.Info("Initialize steps context for evaluating environment url");
+                            context.ExpressionValues["steps"] = context.Global.StepsContext.GetScope(context.ScopeName);
+
+                            var templateEvaluator = context.ToPipelineTemplateEvaluator();
+                            var environmentUrlToken = templateEvaluator.EvaluateEnvironmentUrl(jobContext.ActionsEnvironment.Url, context.ExpressionValues, context.ExpressionFunctions);
```

Comment:
rather than duplicate all of the context setup, can we move that above the block `if (message.JobOutputs != null && message.JobOutputs.Type != TokenType.Null)`